coreos-cloudinit.go

```diff
@@ -216,66 +232,40 @@ func main() {
 		}
 	}
 
-	log.Printf("Fetching meta-data from datasource of type %q\n", ds.Type())
-	metadata, err := ds.FetchMetadata()
+	udata, err := initialize.NewUserData(string(userdataBytes), env)
 	if err != nil {
-		log.Printf("Failed fetching meta-data from datasource: %v\n", err)
-		os.Exit(1)
-	}
-
-	// Apply environment to user-data
-	env := initialize.NewEnvironment("/", ds.ConfigRoot(), flags.workspace, flags.sshKeyName, metadata)
-	userdata := env.Apply(string(userdataBytes))
-
-	var ccu *config.CloudConfig
-	var script *config.Script
-	switch ud, err := initialize.ParseUserData(userdata); err {
-	case initialize.ErrIgnitionConfig:
-		fmt.Printf("Detected an Ignition config. Exiting...")
-		os.Exit(0)
-	case nil:
-		switch t := ud.(type) {
-		case *config.CloudConfig:
-			ccu = t
-		case *config.Script:
-			script = t
-		}
-	default:
-		fmt.Printf("Failed to parse user-data: %v\nContinuing...\n", err)
+		log.Printf("Failed to parse user-data: %v\nContinuing...\n", err)
 		failure = true
 	}
 
-	log.Println("Merging cloud-config from meta-data and user-data")
-	cc := mergeConfigs(ccu, metadata)
+	mustStop := false
+	hostname := determineHostname(metadata, udata)
+	if err := initialize.ApplyHostname(hostname); err != nil {
+		log.Printf("Failed to set hostname: %v", err)
+		mustStop = true
+	}
```

Comment:
This now overwrites any static hostname with the meta-data hostname. Can we remove these lines again?

Reply:
That was kind of the point of this. If coreos-cloudinit is used, it is responsible for setting the hostname. The hostname is fetched from userdata or meta-data and the short form hostname is set. Userdata has precedence.

Reply:
In particular, when Ignition ran and coreos-cloudinit gets triggered through the config drive mechanism, this suddenly overwrites the hostname Ignition set up in `/etc/hostname`.
We could try to skip execution of coreos-cloudinit when Ignition ran but still I wonder if this here wouldn't also cause problems when coreos-cloudinit or some custom image setup was writing to `/etc/hostname` and now this here overwrites it.

Reply:
IMO, we should use coreos-cloudinit in scenarios where it's needed (like OpenStack or wherever we need multipart mime/cloud-config). Otherwise it should be disabled.

If, however it's enabled, we should assume it will overwrite some things set by afterburn.

If custom image setup is needed, that should either be done with coreos-cloudinit via userdata, or it should run after coreos-cloudinit.

Otherwise there is no sane way to have coreos-cloudinit run and be useful. Perhaps this is something that needs to be documented?

Reply:
Then it should probably not run if the userdata is ignition. Is there some other trigger that enables it, besides non-ignition userdata?

If yes, we should probably add flags to disable various bits of it, like setting the hostname and SSH keys (for example).

Reply:
I think flags are a good idea, we also have this for afterburn and it would allow to, e.g., opt out of metadata if it's covered by afterburn for the platform. We should also tweak the units to not have it run at all on certain platforms: e.g., on Digital Ocean it can run twice, once through the regular oem-cloudinit service and once through the configdrive, and it makes sense to disable it for the config drive (this is the case we ran into).

Reply:
Will look into adding those flags as soon as I can.